frontend/storm_wiki/util/search.py

```diff
@@ -0,0 +1,230 @@
+                setattr(
+                    self,
+                    category,
+                    set(id_str.split("_(")[0] for id_str in processed_ids),
+                )
+
+        except Exception as e:
+            logger.error(f"Error in _initialize_domain_restrictions: {e}")
+
+    def _is_valid_wikipedia_source(self, url):
+        if not url:
+            return False
+        parsed_url = urlparse(url)
+        if not parsed_url.netloc:
+            return False
+        domain = parsed_url.netloc.split(".")[-2]
+        combined_set = self.generally_unreliable | self.deprecated | self.blacklisted
+        return (
+            domain not in combined_set or "wikipedia.org" in url
+        )  # Allow Wikipedia URLs
+
+    def forward(
+        self, query_or_queries: Union[str, List[str]], exclude_urls: List[str] = []
+    ) -> List[Dict[str, Any]]:
+        queries = (
+            [query_or_queries]
+            if isinstance(query_or_queries, str)
+            else query_or_queries
+        )
+        all_results = []
+
+        for query in queries:
+            results = self._search_with_fallback(query)
+            all_results.extend(results)
+
+        filtered_results = [
+            r
+            for r in all_results
+            if r["url"] not in exclude_urls
+            and self._is_valid_wikipedia_source(r["url"])
+        ]
+
+        if filtered_results:
+            ranked_results = sorted(
+                filtered_results, key=self._calculate_relevance, reverse=True
+            )
+            return ranked_results[: self.max_results]
+        else:
+            logger.warning(f"No results found for query: {query_or_queries}")
+            return []
+
+    def _search_with_fallback(self, query: str) -> List[Dict[str, Any]]:
+        try:
+            results = self._search(self.primary_engine, query)
+        except Exception as e:
+            logger.warning(
+                f"{self.primary_engine} search failed: {str(e)}. Falling back to {self.fallback_engine}."
+            )
+            if self.fallback_engine:
+                try:
+                    results = self._search(self.fallback_engine, query)
+                except Exception as e:
+                    logger.error(f"{self.fallback_engine} search also failed: {str(e)}")
+                    results = []
+            else:
+                logger.error("No fallback search engine specified or available.")
+                results = []
+
+        return results
+
+    def _search(self, engine: str, query: str) -> List[Dict[str, Any]]:
+        if engine not in self.search_engines:
+            raise ValueError(f"Unsupported or unavailable search engine: {engine}")
+
+        search_engine = self.search_engines[engine]
+        results = search_engine(query)
+
+        logger.info(f"Raw results from {engine}: {results}")
+        return results
+
+    def _search_duckduckgo(self, query: str) -> List[Dict[str, Any]]:
```

Comment:
Could we move this part to `knowledge_storm/rm.py`? Simply setting up a wrapper class for `DuckDuckGoSearchAPIWrapper` would be good. I think people may want to use it but not sure everyone will be able to find it in this frontend folder.